src/types/from_register_request.rs

```diff
@@ -170,58 +179,91 @@ impl From<AudioCodec> for pipeline::AudioCodec {
-            });
-        let preset = match request.encoder_preset.unwrap_or(EncoderPreset::Fast) {
-            EncoderPreset::Ultrafast => encoder::ffmpeg_h264::EncoderPreset::Ultrafast,
-            EncoderPreset::Superfast => encoder::ffmpeg_h264::EncoderPreset::Superfast,
-            EncoderPreset::Veryfast => encoder::ffmpeg_h264::EncoderPreset::Veryfast,
-            EncoderPreset::Faster => encoder::ffmpeg_h264::EncoderPreset::Faster,
-            EncoderPreset::Fast => encoder::ffmpeg_h264::EncoderPreset::Fast,
-            EncoderPreset::Medium => encoder::ffmpeg_h264::EncoderPreset::Medium,
-            EncoderPreset::Slow => encoder::ffmpeg_h264::EncoderPreset::Slow,
-            EncoderPreset::Slower => encoder::ffmpeg_h264::EncoderPreset::Slower,
-            EncoderPreset::Veryslow => encoder::ffmpeg_h264::EncoderPreset::Veryslow,
-            EncoderPreset::Placebo => encoder::ffmpeg_h264::EncoderPreset::Placebo,
+    fn try_from(value: RegisterOutputRequest) -> Result<Self, Self::Error> {
+        const NO_VIDEO_OR_AUDIO: &str =
+            "At least one of \"video\" and \"audio\" fields have to be specified.";
+
+        if value.video.is_none() && value.audio.is_none() {
+            return Err(TypeError::new(NO_VIDEO_OR_AUDIO));
+        }
+        let output_options = value.clone().into();
+        let video = match value.video {
+            Some(v) => Some(pipeline::OutputVideoOptions {
+                initial: v.initial.try_into()?,
+                encoder_opts: pipeline::encoder::EncoderOptions::H264(Options {
+                    preset: v.encoder_preset.into(),
+                    resolution: v.resolution.into(),
+                    output_id: value.output_id.clone().into(),
+                }),
+            }),
+            None => None,
         };
 
-        let encoder_options = encoder::EncoderOptions::H264(encoder::ffmpeg_h264::Options {
-            preset,
-            resolution: request.resolution.into(),
-            output_id: request.output_id.clone().into(),
+        let audio = value.audio.map(|a| pipeline::OutputAudioOptions {
+            initial: a.initial.into(),
+            sample_rate: a.sample_rate,
+            channels: a.channels.into(),
+            forward_error_correction: a.forward_error_correction.unwrap_or(false),
         });
 
-        Ok((
-            request.output_id.into(),
-            pipeline::RegisterOutputOptions {
-                encoder_options,
-                output_options,
-            },
-            request.initial_scene.try_into()?,
-        ))
+        Ok(Self {
+            output_id: value.output_id.into(),
+            output_options,
+            video,
+            audio,
+        })
+    }
+}
+
+impl From<EncoderPreset> for pipeline::encoder::ffmpeg_h264::EncoderPreset {
+    fn from(value: EncoderPreset) -> Self {
+        match value {
+            EncoderPreset::Ultrafast => pipeline::encoder::ffmpeg_h264::EncoderPreset::Ultrafast,
```

Comment:
```suggestion
            EncoderPreset::Ultrafast => encoder::ffmpeg_h264::EncoderPreset::Ultrafast,
```
also this was used only in one place, so moving it to separate `From<>` was not necessary.